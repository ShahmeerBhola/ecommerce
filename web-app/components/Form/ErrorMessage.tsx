import React, { FC } from 'react';

type Props = {
    error: Forms.FormikError;
};

const ErrorMessage: FC<Props> = ({ error }) => {
    return (
        <div className="absolute text-red-100 text-xs">
            <span>{error}</span>
        </div>
    );
};

export default ErrorMessage;
