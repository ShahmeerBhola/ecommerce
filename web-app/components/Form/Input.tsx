import React from 'react';

import Text from '../Text';

type Props = {
    name: string;
    type: string;
    value: string;
    placeholder?: string;
    disabled?: boolean;
    overrideContent?: () => JSX.Element;
    borderColor?: 'border-red-200' | 'border-gray-901';
    onChange?: Forms.InputOnChange;
    onBlur?: Forms.InputOnBlur;
    blankOptionsFallbackText?: string;
    options?: Forms.FieldOption[];
};

const Input: React.FunctionComponent<Props> = ({
    name,
    type,
    value,
    placeholder,
    disabled,
    overrideContent,
    borderColor = 'border-gray-901',
    onChange,
    onBlur,
    blankOptionsFallbackText,
    options,
}) => {
    const baseStyle = 'text-sm text-gray-902 placeholder-gray-902 p-3 focus:outline-none';
    const inputStyle = `${baseStyle} w-full h-50px bg-gray-50 border border-solid ${borderColor}`;
    const textAreaStyle = `${baseStyle} block w-full h-175px bg-gray-50 border border-solid ${borderColor}`;
    const checkboxRadioStyle = `${baseStyle} mr-5`;

    if (type === 'radio') {
        // Radio buttons seem to be a bug in Formik..
        // see here for Github Issue https://github.com/jaredpalmer/formik/issues/1024
        // and here for a solution https://codesandbox.io/s/88jqoy4qvl
        const radioOptions = options as Forms.FieldOption[];

        if (overrideContent) {
            return overrideContent();
        }

        if (radioOptions.length === 0) {
            return <Text>{blankOptionsFallbackText}</Text>;
        }

        return (
            <>
                {placeholder && <Text>{placeholder}</Text>}
                {radioOptions.map(({ value: radioValue, label: radioLabel }) => (
                    <div key={radioValue}>
                        <label htmlFor={`${name}_${radioValue}`}>
                            <input
                                id={`${name}_${radioValue}`}
                                name={name}
                                type="radio"
                                disabled={disabled}
                                value={radioValue}
                                defaultChecked={value === radioValue}
                                onChange={onChange}
                                onBlur={onBlur}
                                className={checkboxRadioStyle}
                            />
                            {radioLabel}
                        </label>
                    </div>
                ))}
            </>
        );
    } else if (type === 'checkbox') {
        // Formik checkboxes are really stubborn...
        // https://github.com/formium/formik/issues/602
        return (
            <>
                <input
                    name={name}
                    type={type}
                    value={value}
                    disabled={disabled}
                    defaultChecked={(value as any) as boolean}
                    className={checkboxRadioStyle}
                    placeholder={placeholder}
                    onChange={onChange}
                    onBlur={onBlur}
                />
                <label htmlFor={name}>{placeholder}</label>
            </>
        );
    } else if (type === 'textarea') {
        return (
            <textarea
                name={name}
                rows={4}
                value={value}
                placeholder={placeholder}
                className={textAreaStyle}
                onChange={onChange}
                onBlur={onBlur}
            />
        );
    }

    return (
        <input
            name={name}
            type={type}
            value={value}
            disabled={disabled}
            className={inputStyle}
            placeholder={placeholder}
            onChange={onChange}
            onBlur={onBlur}
        />
    );
};

export default Input;
