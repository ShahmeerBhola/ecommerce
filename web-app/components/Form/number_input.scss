$input-number-prefix: rc-input-number;

.rc-input-number {
    border: 1px solid #d9d9d9;
    border-radius: 4px;
    display: inline-block;
    font-size: 12px;
    height: 26px;
    line-height: 26px;
    margin: 0;
    padding: 0;
    transition: all 3s;
    vertical-align: middle;

    &-focused {
        border-color: #1890ff;
        box-shadow: 0 0 0 2px rgba(24, 144, 255, 0.2);
    }

    &-handler {
        display: block;
        height: 12px;
        line-height: 12px;
        overflow: hidden;
        text-align: center;
        touch-action: none;

        &-active {
            background: #ddd;
        }
    }

    &-handler-up-inner,
    &-handler-down-inner {
        color: #666;
        user-select: none;
    }

    &:hover,
    &:focus {
        border-color: #1890ff;

        .#{input-number-prefix}-handler-up,
        .#{input-number-prefix}-handler-wrap {
            border-color: #1890ff;
        }
    }

    &-disabled:hover,
    &-disabled:focus {
        border-color: #d9d9d9;

        .#{input-number-prefix}-handler-up,
        .#{input-number-prefix}-handler-wrap {
            border-color: #d9d9d9;
        }
    }

    &-input-wrap {
        height: 100%;
        overflow: hidden;
    }

    &-input {
        appearance: textfield;
        border: 0;
        border-radius: 4px;
        color: #666;
        height: 100%;
        line-height: 26px;
        outline: 0;
        padding: 0;
        text-align: center;
        transition: all 0.3s ease;
        width: 100%;
    }

    &-handler-wrap {
        border-left: 1px solid #d9d9d9;
        float: right;
        height: 100%;
        transition: all 3s;
        width: 20px;
    }

    &-handler-up {
        border-bottom: 1px solid #d9d9d9;
        padding-top: 1px;
        transition: all 3s;

        &-inner {
            &::after {
                content: '+';
            }
        }
    }

    &-handler-down {
        transition: all 3s;

        &-inner {
            &::after {
                content: '-';
            }
        }
    }

    @mixin handler-disabled() {
        opacity: 0.72;

        &:hover,
        &:focus {
            border-color: #d9d9d9;
            color: #999;
        }
    }

    &-handler-down-disabled,
    &-handler-up-disabled {
        @include handler-disabled();
    }

    &-disabled {
        .#{input-number-prefix}-input {
            background-color: #f3f3f3;
            cursor: not-allowed;
            opacity: 0.72;
        }

        .#{input-number-prefix}-handler {
            @include handler-disabled();
        }
    }
}
