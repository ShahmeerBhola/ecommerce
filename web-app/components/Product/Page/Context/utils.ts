import isEmpty from 'lodash/isEmpty';

import type { TProductContext } from './types';

export const getSingleAttributes = (_attributes?: StringMap<string[]>): StringMap => {
    let singleAttributes = {};

    for (const [key, value] of Object.entries(_attributes || {})) {
        if (value.length === 1) {
            singleAttributes = {
                ...singleAttributes,
                [key]: value[0],
            };
        }
    }

    return singleAttributes;
};

export const getAvailableAttributes = (
    variations: WordPress.ProductVariation[],
    newAttributes: StringMap,
): StringMap<string[]> => {
    const availableAttribute: StringMap<string[]> = {};

    for (const { attributes } of variations) {
        let isAvailableAttribute = true;

        for (const [key, value] of Object.entries(newAttributes)) {
            isAvailableAttribute = isAvailableAttribute && attributes[key] === value;
        }

        if (isAvailableAttribute) {
            for (const [key, value] of Object.entries(attributes)) {
                const currentAttribute = availableAttribute[key] || [];

                if (currentAttribute.indexOf(value) > -1) {
                    continue;
                }

                availableAttribute[key] = [...currentAttribute, value];
            }
        }
    }

    return availableAttribute;
};

const getAttributes = (
    variations: WordPress.ProductVariation[],
    selectedAttributes: {} | StringMap,
): TProductContext.CheckAttributesType => {
    const availableAttributes = getAvailableAttributes(variations, selectedAttributes);
    const singleAttributes = getSingleAttributes(availableAttributes);

    return {
        currentAttributes: {
            ...selectedAttributes,
            ...singleAttributes,
        },
        availableAttributes,
    };
};

export const checkAttributes = (state: TProductContext.State, payload: any): TProductContext.CheckAttributesType => {
    const { currentAttributes, product } = state;
    const { variations } = product as WordPress.Product;
    const newAttributes = {
        ...currentAttributes,
        ...payload,
    };

    const attributes = getAttributes(variations, newAttributes);

    if (isEmpty(attributes.availableAttributes)) {
        const resetAttributes = {
            ...payload,
        };

        return getAttributes(variations, resetAttributes);
    }

    return attributes;
};

export const isActiveAttribute = (attributes: StringMap, currentAttributes: StringMap): boolean => {
    return Object.entries(attributes).reduce<boolean>((isSelectedAttribute, attribute) => {
        const [key, value] = attribute;
        const matchedAttribute = value === '' || currentAttributes[key] === value;

        return isSelectedAttribute && matchedAttribute;
    }, true);
};
