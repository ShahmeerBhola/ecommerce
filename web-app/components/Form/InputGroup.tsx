import React, { FC } from 'react';

import ErrorMessage from './ErrorMessage';

type Props = {
    touched?: boolean;
    className?: string;
    error?: Forms.FormikError;
    noBottomPadding?: boolean;
    noXPadding?: boolean;
};

const InputGroup: FC<Props> = ({
    error,
    touched,
    className,
    noBottomPadding = false,
    noXPadding = false,
    children,
}) => {
    const isError = error && touched;
    const classNames = `pt-2 box-border ${noXPadding ? '' : 'px-2'} ${className}`;

    return (
        <div className={`${classNames} ${noBottomPadding ? '' : 'pb-3'}`}>
            <div className="relative h-full">
                {children}
                {isError && <ErrorMessage error={error} />}
            </div>
        </div>
    );
};

export default InputGroup;
