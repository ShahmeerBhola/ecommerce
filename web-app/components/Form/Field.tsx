import React from 'react';
import { CardElement } from '@stripe/react-stripe-js';
import { Field, FieldProps } from 'formik';

import { formatPhoneNumber } from '../../utils';

import Select from './Select';
import Input from './Input';
import InputGroup from './InputGroup';
import CreditCardLogos from '../CreditCardLogos';

import './stripe_card_element.scss';

type StripeCallbackProps = {
    error: Forms.FormikError;
    touched: boolean;
    updateError: (err: Forms.FormikError) => void;
    updateValid: (valid: boolean) => void;
    updateTouched: (touched: boolean) => void;
};

interface Props<IV> extends Partial<StripeCallbackProps> {
    name: string;
    type: string;
    placeholder?: string;
    className?: string;
    disabled?: boolean;
    noXPadding?: boolean;
    overrideContent?: () => JSX.Element;
    blankOptionsFallbackText?: string;
    onChange?: Forms.ChangeHandler<IV>;
    onBlur?: Forms.ChangeHandler<IV>;
    fieldOptions?: Forms.FieldOption[];
}

const SOSField = <IV,>({
    name,
    type,
    placeholder,
    className,
    onChange,
    onBlur,
    noXPadding,
    fieldOptions,
    disabled,
    overrideContent,
    blankOptionsFallbackText,
    ...rest
}: Props<IV>): JSX.Element => {
    return (
        <Field name={name}>
            {({
                field: { name, value: fieldValue },
                form: { handleBlur, setFieldValue, values },
                meta: { error, touched },
            }: FieldProps<IV>): JSX.Element => {
                const borderColor = error && touched ? 'border-red-200' : 'border-gray-901';

                const value = fieldValue as any; // eslint-disable-line @typescript-eslint/no-explicit-any
                const onChangeWrapper = (e: Forms.ChangeEvent): void => {
                    let newValue: any = e.currentTarget.value; // eslint-disable-line @typescript-eslint/no-explicit-any

                    // Do any normalization here..
                    if (type === 'tel') {
                        newValue = formatPhoneNumber(newValue);
                    } else if (type === 'checkbox') {
                        newValue = (e as React.ChangeEvent<HTMLInputElement>).currentTarget.checked;
                    }

                    setFieldValue(name, newValue);
                    onChange && onChange(newValue, values);
                };

                const onBlurWrapper = (e: Forms.FocusEvent): void => {
                    let newValue: any; // eslint-disable-line @typescript-eslint/no-explicit-any

                    if (type === 'checkbox') {
                        newValue = (e as React.ChangeEvent<HTMLInputElement>).currentTarget.checked;
                    } else {
                        newValue = e.currentTarget.value;
                    }

                    handleBlur(e);
                    onBlur && onBlur(newValue, values);
                };

                if (type === 'select') {
                    return (
                        <InputGroup error={error} touched={touched} className={className} noXPadding={noXPadding}>
                            <Select
                                name={name}
                                value={value}
                                placeholder={placeholder}
                                disabled={disabled}
                                onChange={onChangeWrapper}
                                onBlur={onBlurWrapper}
                                options={fieldOptions as Forms.FieldOption[]}
                            />
                        </InputGroup>
                    );
                } else if (type === 'stripe') {
                    // The reason why we have to have error, touched, updateError, updateTouched
                    // and updateValid is that the Stripe CardElement component does not expose its value
                    // therefore it would be impossible to do form validation on this field (ex. requiring that the
                    // field has a value). Instead we store the state for this field, error and touched, outside
                    // of our form and pass it all the way down to the field. The three update methods simply
                    // update that state
                    const {
                        error: stripeError,
                        touched: stripeTouched,
                        updateError,
                        updateTouched,
                        updateValid,
                    } = rest as StripeCallbackProps;
                    const stripeRed = 'rgb(242, 57, 57)';
                    const stripeGray = 'rgb(124, 129, 134)';

                    return (
                        <InputGroup
                            error={stripeError}
                            touched={stripeTouched}
                            className={className}
                            noXPadding={noXPadding}
                        >
                            <CardElement
                                className={`sos-stripe-card-element${stripeError !== null ? ' error' : ''}`}
                                options={{
                                    hidePostalCode: true,
                                    style: {
                                        base: {
                                            fontFamily: 'Poppins, sans-serif',
                                            fontSize: '16px',
                                            color: stripeGray,
                                            '::placeholder': {
                                                color: stripeGray,
                                            },
                                        },
                                        invalid: {
                                            color: stripeRed,
                                        },
                                    },
                                }}
                                onChange={({ error, complete }): void => {
                                    updateTouched(true);

                                    if (error?.message) {
                                        updateValid(false);
                                        updateError(error.message);
                                    } else {
                                        updateError(null);
                                    }

                                    if (!error && complete) {
                                        updateValid(true);
                                    }
                                }}
                            />
                            <CreditCardLogos height="h-30px" className="mt-3" />
                        </InputGroup>
                    );
                }

                return (
                    <InputGroup error={error} touched={touched} className={className} noXPadding={noXPadding}>
                        <Input
                            name={name}
                            type={type}
                            value={value}
                            disabled={disabled}
                            overrideContent={overrideContent}
                            borderColor={borderColor}
                            placeholder={placeholder}
                            onChange={onChangeWrapper}
                            onBlur={onBlurWrapper}
                            blankOptionsFallbackText={blankOptionsFallbackText}
                            options={fieldOptions}
                        />
                    </InputGroup>
                );
            }}
        </Field>
    );
};

export default SOSField;
