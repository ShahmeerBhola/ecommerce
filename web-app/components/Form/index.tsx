import React from 'react';
import { Formik, FormikProps, FormikHelpers } from 'formik';
import { useToasts } from 'react-toast-notifications';

import { sendSentryEvent, shouldHideFormSection } from '../../utils';

import Section from './Section';
import Submit from './Submit';
import InputGroup from './InputGroup';

type FormSubmitProps = {
    isSubmitting: boolean;
    isValid: boolean;
    noXPadding?: boolean;
};

type FormProps<IV> = {
    initialValues: IV;
    form: Forms.FormDefinition<IV>;
    validationSchema: Forms.YupValidationSchema<IV> | Forms.YupLazyValidationSchema<IV>;
    onSubmit: (data: IV, actions: FormikHelpers<IV>) => void | Promise<void>;
    isInitialValid?: boolean;
    onChangeHandlers?: Forms.ChangeHandlers<IV>;
    onBlurHandlers?: Forms.ChangeHandlers<IV>;
    fieldOptions?: Forms.FieldOptions<IV>;
    hiddenFormSectionConditions?: Forms.HiddenFormSectionCondition<IV>[];
    disabledFields?: Forms.DisabledFields<IV>;
    overriddenFields?: Forms.OverriddenFields<IV>;
    externalDisableSubmit?: boolean;
    additionalExternalSections?: Forms.FormSection<IV>[];
};

const Form = <IV,>(props: FormProps<IV>): JSX.Element => {
    const {
        initialValues,
        validationSchema,
        isInitialValid = true,
        onSubmit,
        form: { sections, submit, renderSubmitInsideSections = true },
        externalDisableSubmit = false,
        additionalExternalSections,
        hiddenFormSectionConditions = [],
        ...rest
    } = props;
    const { addToast } = useToasts();

    const FormSubmit: React.FunctionComponent<FormSubmitProps> = ({ isSubmitting, isValid, noXPadding }) => {
        const { className, loadingMessage, text, noBottomPadding } = submit;
        return (
            <InputGroup className={className} noBottomPadding={noBottomPadding} noXPadding={noXPadding}>
                <Submit
                    className={className}
                    isSubmitting={isSubmitting}
                    disabled={externalDisableSubmit || isSubmitting || !isValid}
                    loadingMessage={loadingMessage}
                >
                    {text}
                </Submit>
            </InputGroup>
        );
    };

    return (
        <Formik<IV>
            isInitialValid={isInitialValid}
            initialValues={initialValues}
            validationSchema={validationSchema}
            enableReinitialize={true}
            onSubmit={(values, helpers): void => {
                try {
                    onSubmit(values, helpers);
                } catch (err) {
                    addToast('Oops, there was an unknown error. Try again', { appearance: 'error' });
                    sendSentryEvent('Unknown form submission exception', { ...props, values });
                    helpers.setSubmitting(false);
                }
            }}
        >
            {({ values, isSubmitting, isValid, handleSubmit }: FormikProps<IV>): JSX.Element => (
                <form onSubmit={handleSubmit}>
                    {sections.map(({ id, label, fields, noXPadding, ...sectionRest }, idx) => (
                        <Section<IV>
                            key={`${label}-${idx}`}
                            label={label}
                            hidden={shouldHideFormSection<IV>(hiddenFormSectionConditions, values, id)}
                            fields={fields}
                            noXPadding={noXPadding}
                            {...sectionRest}
                            {...rest}
                        >
                            {renderSubmitInsideSections && idx === sections.length - 1 ? (
                                <FormSubmit isSubmitting={isSubmitting} isValid={isValid} noXPadding={noXPadding} />
                            ) : null}
                        </Section>
                    ))}
                    {additionalExternalSections?.map(({ id, label, fields, ...sectionRest }, idx) => (
                        <Section<IV>
                            key={`${label}-${idx}`}
                            label={label}
                            fields={fields}
                            hidden={shouldHideFormSection<IV>(hiddenFormSectionConditions, values, id)}
                            {...sectionRest}
                            {...rest}
                        />
                    ))}
                    {!renderSubmitInsideSections ? <FormSubmit isSubmitting={isSubmitting} isValid={isValid} /> : null}
                </form>
            )}
        </Formik>
    );
};

export default Form;
