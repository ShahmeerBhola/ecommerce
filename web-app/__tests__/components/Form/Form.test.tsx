import React from 'react';
import renderer from 'react-test-renderer';
import Form from '../../../components/Form';
import { ToastProvider } from '../../utils';

describe('Form component', () => {
  test('it renders properly', () => {
    const initialValues = {
      name: '',
    };

    const tree = renderer
      .create(
        <ToastProvider>
          <Form<typeof initialValues>
            form={{
              sections: [
                {
                  id: 'contact_information',
                  label: 'Yoooo dog',
                  fields: [],
                },
              ],
              submit: {
                className: 'submit-field-class-name',
                text: 'Submit that Form',
                loadingMessage: 'Non default loading message',
              },
            }}
            validationSchema={(): void => {
              console.log('validation schema');
            }}
            initialValues={initialValues}
            onSubmit={(): void => {
              console.log('yo');
            }}
          />
        </ToastProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
