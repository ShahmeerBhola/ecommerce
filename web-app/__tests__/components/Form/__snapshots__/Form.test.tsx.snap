// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component it renders properly 1`] = `
Array [
  <form
    className="font-secondary"
    onSubmit={[Function]}
  >
    <div
      className="flex flex-wrap my-3"
    >
      <p
        className="font-secondary font-normal text-gray-700 text-lg w-full mb-3 font-medium"
      >
        Yoooo dog
      </p>
      <div
        className="pt-2 px-2 box-border pb-5 submit-field-class-name"
      >
        <button
          className="font-secondary font-bold py-2 px-5 uppercase disabled:opacity-75 disabled:cursor-not-allowed focus:outline-none h-full w-full text-sm text-white bg-red-100 hover:bg-red-200"
          disabled={false}
          type="submit"
        >
          Submit that Form
        </button>
      </div>
    </div>
  </form>,
  <div
    className="react-toast-notifications__container css-mrh2d-ToastContainer"
  />,
]
`;
